notify/notify_test.go

```diff
@@ -166,10 +170,78 @@ func TestDedupingNotifierHasUpdate(t *testing.T) {
 	}
 }
 
-func TestDedupingNewNotifiesExtraction(t *testing.T) {
+func TestMultiStage(t *testing.T) {
+	ctx := context.Background()
+	now := time.Now()
+
+	alerts := []*types.Alert{
+		{
+			Alert: model.Alert{
+				Labels: model.LabelSet{"alertname": "0"},
+			},
+		}, {
+			Alert: model.Alert{
+				Labels: model.LabelSet{"alertname": "1"},
+				EndsAt: now.Add(-5 * time.Minute),
+			},
+		},
+	}
+
+	s1 := equalityStage{alerts: alerts}
+	s2 := equalityStage{alerts: alerts}
+	stage := MultiStage{s1, s2}
+
+	out, err := stage.Exec(ctx, alerts...)
+	if err != nil {
+		t.Fatalf("Exec failed: %s", err)
+	}
+
+	if !reflect.DeepEqual(alerts, s1.alerts) ||
+		!reflect.DeepEqual(s1.alerts, s2.alerts) ||
+		!reflect.DeepEqual(s2.alerts, out) {
```

Comment:
Bit confused on this one. Aren't `alerts`, `s1.alerts`, and `s2.alerts` all static and not changed in `equalityStage` either? Why the comparisons?

I think what we really want to check here is a) that the stages are evaluated in the correct order, and b) that the output of the previous one is the input of the next one.

Maybe we can construct a `type StageFunc func(context.Context, ...*types.Alert) ([]*types.Alert, error)` that allows us to define new mock stages inline in this test.

Something like 

```go
var (
    alerts1 = []*types.Alert{{}}
    alerts2 = []*types.Alert{{}, {}}
    alerts3 = []*types.Alert{{}, {}, {}}
)

stage := MultiStage{
    StageFunc(func(context.Context, alerts ...*types.Alert) ([]*types.Alert, error) {
        if !equal(alerts, alerts1) {
            // something went wrong
        }
        return alerts2
    }),
}